File: benchmarks/pandas/bench_resample_label_closed.py
"""
Benchmark: resample_label_closed — pd.Series.resample with explicit label options.

Mirrors tsb: resampleSeries(s, "H", { label: "right" }).sum() and
             resampleSeries(s, "H", { label: "left" }).mean()

Uses a 50k-row minute-resolution dataset.

Outputs JSON: {"function": "resample_label_closed", "mean_ms": ..., "iterations": ..., "total_ms": ...}
"""
import json
import time
import math
import numpy as np
import pandas as pd

SIZE = 50_000
WARMUP = 3
ITERATIONS = 30

base = pd.Timestamp("2020-01-01T00:00:00Z")
idx = pd.date_range(base, periods=SIZE, freq="min")
data = np.array([(i % 200) * 0.5 + math.sin(i * 0.02) * 20 for i in range(SIZE)])
s = pd.Series(data, index=idx)

# Warm up
for _ in range(WARMUP):
    s.resample("H", label="right").sum()
    s.resample("H", label="left").mean()

times = []
for _ in range(ITERATIONS):
    t0 = time.perf_counter()
    s.resample("H", label="right").sum()
    s.resample("H", label="left").mean()
    times.append((time.perf_counter() - t0) * 1000)

total = sum(times)
print(json.dumps({
    "function": "resample_label_closed",
    "mean_ms": total / ITERATIONS,
    "iterations": ITERATIONS,
    "total_ms": total,
}))

File: benchmarks/tsb/bench_resample_label_closed.ts
/**
 * Benchmark: resample_label_closed — SeriesResampler with explicit label options.
 *
 * Mirrors pandas: pd.Series.resample("H", label="right").sum() and
 *                 pd.Series.resample("H", label="left").mean()
 *
 * Uses a 50k-row minute-resolution dataset.  Benchmarks the non-default label
 * path where keyToLabel must convert group keys (e.g. label="right" on an
 * hourly series whose default label is "left").
 *
 * Outputs JSON: {"function": "resample_label_closed", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { Series, resampleSeries } from "../../src/index.ts";

const SIZE = 50_000;
const WARMUP = 3;
const ITERATIONS = 30;

const base = new Date("2020-01-01T00:00:00Z").getTime();
const idx = Array.from({ length: SIZE }, (_, i) => new Date(base + i * 60_000));
const data = Array.from({ length: SIZE }, (_, i) => (i % 200) * 0.5 + Math.sin(i * 0.02) * 20);

const s = new Series({ data, index: idx });

// Warm up
for (let i = 0; i < WARMUP; i++) {
  resampleSeries(s, "H", { label: "right" }).sum();
  resampleSeries(s, "H", { label: "left" }).mean();
}

const times: number[] = [];
for (let i = 0; i < ITERATIONS; i++) {
  const t0 = performance.now();
  resampleSeries(s, "H", { label: "right" }).sum();
  resampleSeries(s, "H", { label: "left" }).mean();
  times.push(performance.now() - t0);
}
const total = times.reduce((a, b) => a + b, 0);
console.log(
  JSON.stringify({
    function: "resample_label_closed",
    mean_ms: total / ITERATIONS,
    iterations: ITERATIONS,
    total_ms: total,
  }),
);
